core/record/recordExpressionParser.js

```diff
@@ -8,17 +8,23 @@ const NodeDefExpression = require('@core/survey/nodeDefExpression')
 const Record = require('./record')
 const Node = require('./node')
 const Expression = require('@core/expressionParser/expression')
+const SystemError = require('@server/utils/systemError')
+const Validation = require('@core/validation/validation')
 
 const evalNodeQuery = async (survey, record, node, query) => {
   const ctx = {
     node: bindNode(survey, record, node),
-    functions: {
-      [Expression.types.ThisExpression]: (expr, { node }) => node
-    },
   }
   return await Expression.evalString(query, ctx)
 }
 
+// Get reachable nodes, i.e. the children of the node's ancestors.
+// NOTE: The root node is excluded, but it _should_ be an entity, so that is fine.
+const getReachableNodes = (record, node) => {
+  const ancestors = Record.getAncestorsAndSelf(node)(record)
+  return R.flatten(ancestors.map(n => Record.getNodeChildren(n)(record)))
+}
```

Comment:
Hm I was thinking that the data is already in memory, so it's not more than a few thousand pointers at the most?
That seems manageable to me. Maybe I'm missing something?

Reply:
Yes, this would work when editing  single record, but if run this for N records during survey publish it might get slow...

Reply:
The tradeoff here is that for 10-20kB more RAM allocation for thousands of pointers, the code is (IMO) simpler.

I could benchmark this, but I suspect that the difference would be less than a millisecond vs. a version that avoid allocating an array, in the case of a survey with thousands of nodes.

Does that match what you had in mind?

Reply:
This can be achieved with a completely of O(1). I thinking with the hierarchy uuids with store it just simplify this. Take a look at pull request. Let us know what you think